# Лабораторная работа № 1

# Работа с библиотекой pandas

**Цель работы:** познакомиться с использованием библиотеки pandas при решении практических задач, связанных с анализом данных.

Проанализируем набор данных об автомобилях. Датасет содержит сведения о характеристиках автомобилей. Также он содержит информацию о различных параметрах, используемых при анализе автомобилей. Смысл задачи – выполнять обработку, объединение и анализ данных, основываясь на этих данных. Набор данных включает следующие атрибуты:
- **mpg** – расход топлива (miles per gallon);
- **cylinders** – цилиндры;
- **displacement** – пробег;
- **horsepower** – мощность двигателя в л.с.;
- **weight** – вес;
- **acceleration** – ускорение;
- **model** – год выпуска;
- **origin** – происхождение;
- **car** – название модели.

# 0. Импортируем необходимые библиотеки

In [1]:
import numpy as np
import pandas as pd

# 1. Загружаем датасеты auto1.csv, auto2.csv и помещаем их в датафреймы

In [2]:
auto1 = pd.read_csv('auto1.csv')
auto2 = pd.read_csv('auto2.csv')

# 2. В наборе auto1 есть безымянные пустые столбцы, исправим это

In [3]:
auto1 = auto1.loc[:, ~auto1.columns.str.contains('^Unnamed')]

# 3. Объединим наборы auto1 и auto2 в один датасет auto

In [4]:
auto = pd.concat([auto1, auto2], ignore_index=True)

# 4. Выведем 7 первых и 7 последних строк датафрейма auto

In [5]:
print(auto.head(7))
print()
print(auto.tail(7))

    mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0  18.0          8           307        130    3504          12.0     70   
1  15.0          8           350        165    3693          11.5     70   
2  18.0          8           318        150    3436          11.0     70   
3  16.0          8           304        150    3433          12.0     70   
4  17.0          8           302        140    3449          10.5     70   
5  15.0          8           429        198    4341          10.0     70   
6  14.0          8           454        220    4354           9.0     70   

   origin                        car  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino  
5       1           ford galaxie 500  
6       1           chevrolet impala  

      mpg  cylinders  displacement horsepower  weight  acceleration  model  \


# 5. Получим общую информацию о датасете с помощью метода info()

In [6]:
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    int64  
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model         398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car           398 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 28.1+ KB


# 6. Выведем с помощью метода describe() основные статистические показатели для всех числовых признаков

In [7]:
auto.describe()

,mpg,cylinders,displacement,weight,acceleration,model,origin
count,398.000000,398.000000,398.000000,398.000000,398.000000,398.000000,398.000000
mean,23.514573,5.454774,193.427136,2970.424623,15.568090,76.010050,1.572864
std,7.815984,1.701004,104.268683,846.841774,2.757689,3.697627,0.802055
min,9.000000,3.000000,68.000000,1613.000000,8.000000,70.000000,1.000000
25%,17.500000,4.000000,104.250000,2223.750000,13.825000,73.000000,1.000000
50%,23.000000,4.000000,148.500000,2803.500000,15.500000,76.000000,1.000000
75%,29.000000,8.000000,262.000000,3608.000000,17.175000,79.000000,2.000000
max,46.600000,8.000000,455.000000,5140.000000,24.800000,82.000000,3.000000


> По результатам метода describe() видно, что для всех числовых признаков количество наблюдений равно 398. Среднее значение признака mpg составляет 23.51, среднее число цилиндров — 5.45, а средний вес автомобиля — 2970.42. Значения модельного года находятся в диапазоне от 70 до 82.

# 7. Для категориального признака car определим количество уникальных категорий и частоту встречаемости каждой из них

In [8]:
print('Number of unique categories: ', auto['car'].nunique())
print()
print(auto['car'].value_counts().to_string())

Number of unique categories:  305

car
ford pinto                              6
toyota corolla                          5
amc matador                             5
ford maverick                           5
chevrolet chevette                      4
amc gremlin                             4
chevrolet impala                        4
peugeot 504                             4
amc hornet                              4
toyota corona                           4
ford gran torino                        3
chevrolet caprice classic               3
honda civic                             3
volkswagen dasher                       3
dodge colt                              3
chevrolet citation                      3
chevrolet nova                          3
pontiac catalina                        3
plymouth fury iii                       3
ford galaxie 500                        3
chevrolet vega                          3
plymouth duster                         3
amc matador (sw)                     

# 8. Используем группировку и агрегирующие функции

Преобразуем признак horsepower к числовому типу для дальнейших вычислений

In [9]:
auto['horsepower'] = pd.to_numeric(auto['horsepower'], errors='coerce')

Определим средний расход топлива, мощность и вес у автомобилей из каждого региона происхождения

In [10]:
stats_by_origin = auto.groupby('origin')[['mpg', 'horsepower', 'weight']].mean()
print('Средние показатели по регионам происхождения:')
print(stats_by_origin)

Средние показатели по регионам происхождения:
              mpg  horsepower       weight
origin                                    
1       20.083534  119.048980  3361.931727
2       27.891429   80.558824  2423.300000
3       30.450633   79.835443  2221.227848


> По результатам группировки по признаку origin видно, что автомобили из региона 3 имеют наибольший средний расход топлива (30.45) и наименьший средний вес (2221.23). Автомобили из региона 1, наоборот, имеют наименьший средний расход топлива (20.08), но при этом наибольшую среднюю мощность (119.05) и наибольший средний вес (3361.93).

Определим, как количество цилиндров влияет на среднюю мощность, ускорение и расход топлива

In [11]:
stats_by_cylinders = auto.groupby('cylinders')[['horsepower', 'acceleration', 'mpg']].mean()
print('Средние показатели в зависимости от количества цилиндров:')
print(stats_by_cylinders)

Средние показатели в зависимости от количества цилиндров:
           horsepower  acceleration        mpg
cylinders                                     
3           99.250000     13.250000  20.550000
4           78.281407     16.601471  29.286765
5           82.333333     18.633333  27.366667
6          101.506024     16.263095  19.985714
8          158.300971     12.955340  14.963107


> По результатам группировки по количеству цилиндров видно, что автомобили с большим числом цилиндров в среднем обладают более высокой мощностью и меньшим значением mpg. Наибольшая средняя мощность наблюдается у автомобилей с 8 цилиндрами (158.30), при этом у них же одно из наименьших значений mpg (14.96). Автомобили с 4 цилиндрами, напротив, имеют меньшую среднюю мощность (78.28) и более высокое значение mpg (29.29).

Определим, как менялись средняя мощность двигателей и средний вес автомобилей с течением времени

In [12]:
stats_by_model = auto.groupby('model')[['horsepower', 'weight']].mean()
print('Средняя мощность и средний вес по годам:')
print(stats_by_model)

Средняя мощность и средний вес по годам:
       horsepower       weight
model                         
70     147.827586  3372.793103
71     107.037037  2995.428571
72     120.178571  3237.714286
73     130.475000  3419.025000
74      94.230769  2877.925926
75     101.066667  3176.800000
76     101.117647  3078.735294
77     105.071429  2997.357143
78      99.694444  2861.805556
79     101.206897  3055.344828
80      77.481481  2436.655172
81      81.035714  2522.931034
82      81.466667  2453.548387


> По результатам группировки по модельному году видно, что в более поздние годы средние значения мощности двигателей и веса автомобилей в целом становятся ниже по сравнению с началом рассматриваемого периода.

# 9. Используем агрегирующие функции agg()

Определим для автомобилей каждого региона средние, минимальные и максимальные значения расхода топлива, медианный вес и стандартное отклонение по мощности

In [13]:
stats_region_agg = auto.groupby('origin').agg({
    'mpg': ['mean', 'min', 'max'],
    'weight': 'median',
    'horsepower': 'std'
})

print('Агрегированные показатели по регионам:')
print(stats_region_agg)

Агрегированные показатели по регионам:
              mpg              weight horsepower
             mean   min   max  median        std
origin                                          
1       20.083534   9.0  39.0  3365.0  39.897790
2       27.891429  16.2  44.3  2240.0  20.157871
3       30.450633  18.0  46.6  2155.0  17.819199


> По результатам агрегирования по регионам видно, что автомобили из региона 3 имеют наибольшее среднее значение mpg (30.45), а также максимальное значение mpg, равное 46.6. Автомобили из региона 1 имеют наибольший медианный вес (3365.0) и наибольшее стандартное отклонение мощности (39.90), что говорит о большем разбросе значений мощности в этом регионе.

Определим по каждому году выпуска размах ускорения и суммарную мощность для всех выпущенных в этом году машин

In [14]:
stats_model_agg = auto.groupby('model').agg({
    'acceleration': lambda x: x.max() - x.min(),
    'horsepower': 'sum'
})

print('Размах ускорения и суммарная мощность по годам:')
print(stats_model_agg)

Размах ускорения и суммарная мощность по годам:
       acceleration  horsepower
model                          
70             12.5      4287.0
71              9.0      2890.0
72             12.5      3365.0
73             11.5      5219.0
74              7.5      2450.0
75              9.5      3032.0
76             10.2      3438.0
77              7.9      2942.0
78             10.3      3589.0
79             13.5      2935.0
80             12.3      2092.0
81              8.1      2269.0
82             13.0      2444.0


> По результатам агрегирования по модельному году видно, что наибольший размах ускорения наблюдается у автомобилей 79 и 82 годов выпуска (13.5 и 13.0 соответственно). Наибольшая суммарная мощность приходится на 73 год выпуска (5219.0), что означает, что в этом году суммарная мощность всех представленных автомобилей была максимальной.

# 10. Используем сводные таблицы pivot_table()

Определим среднюю экономичность автомобилей в зависимости от года выпуска и региона происхождения

In [15]:
pivot_mpg = pd.pivot_table(
    auto,
    values='mpg',
    index='model',
    columns='origin',
    aggfunc='mean'
)

print('Средняя экономичность автомобилей по годам и регионам:')
print(pivot_mpg)

Средняя экономичность автомобилей по годам и регионам:
origin          1          2          3
model                                  
70      15.272727  25.200000  25.500000
71      18.100000  28.750000  29.500000
72      16.277778  22.000000  24.200000
73      15.034483  24.000000  20.000000
74      18.333333  27.000000  29.333333
75      17.550000  24.500000  27.500000
76      19.431818  24.250000  28.000000
77      20.722222  29.250000  27.416667
78      21.772727  24.950000  29.687500
79      23.478261  30.450000  32.950000
80      25.914286  37.288889  35.400000
81      27.530769  31.575000  32.958333
82      29.450000  40.000000  34.888889


> По сводной таблице видно, что в более поздние годы выпуска среднее значение mpg в целом увеличивается для всех регионов происхождения. Наибольшие значения экономичности наблюдаются у автомобилей из регионов 2 и 3 в последние годы рассматриваемого периода.

Создадим сводную таблицу, где для каждого года выпуска и региона происхождения одновременно отображены средний расход топлива и среднее ускорение

In [16]:
pivot_mpg_acc = pd.pivot_table(
    auto,
    values=['mpg', 'acceleration'],
    index='model',
    columns='origin',
    aggfunc='mean'
)

print('Сводная таблица по годам и регионам:')
print(pivot_mpg_acc)

Сводная таблица по годам и регионам:
       acceleration                              mpg                      
origin            1          2          3          1          2          3
model                                                                     
70        11.977273  16.500000  14.750000  15.272727  25.200000  25.500000
71        14.575000  16.750000  16.375000  18.100000  28.750000  29.500000
72        14.055556  18.700000  15.400000  16.277778  22.000000  24.200000
73        13.620690  16.428571  15.625000  15.034483  24.000000  20.000000
74        15.966667  15.333333  17.666667  18.333333  27.000000  29.333333
75        16.350000  15.083333  16.000000  17.550000  24.500000  27.500000
76        15.786364  16.050000  16.575000  19.431818  24.250000  28.000000
77        15.238889  15.000000  16.316667  20.722222  29.250000  27.416667
78        15.545455  16.233333  16.200000  21.772727  24.950000  29.687500
79        15.243478  18.400000  17.200000  23.478261  30.450000

> По сводной таблице видно, что для каждого года выпуска и региона происхождения одновременно представлены средние значения расхода топлива и ускорения. Это позволяет сравнивать автомобили разных регионов не только по экономичности, но и по динамическим характеристикам.

# 11. Создадим серию случайных чисел от 10 000 до 60 000 и добавим столбец owners в датафрейм auto

In [17]:
auto['owners'] = np.random.randint(10000, 60001, size=len(auto))

# 12. Выведем полученный датафрейм на диск в виде CSV-файла

In [18]:
auto.to_csv('auto_result.csv', index=False)